README.md
# node-red-contrib-tplink

TP-Link Smart Home Node-Red Nodes

[![GitHub release](https://img.shields.io/github/release/caseyjhol/node-red-contrib-tplink.svg?style=flat-square)](https://github.com/caseyjhol/node-red-contrib-tplink/releases) [![NPM Version](https://img.shields.io/npm/v/node-red-contrib-tplink.svg?style=flat-square)](https://www.npmjs.com/package/node-red-contrib-tplink) [![GitHub last commit](https://img.shields.io/github/last-commit/caseyjhol/node-red-contrib-tplink.svg?style=flat-square)](https://github.com/caseyjhol/node-red-contrib-tplink/commits/master)

[![Node version](https://img.shields.io/node/v/node-red-contrib-tplink.svg?style=flat-square)](http://nodejs.org/download/) [![GitHub repo size in bytes](https://img.shields.io/github/repo-size/caseyjhol/node-red-contrib-tplink.svg?style=flat-square)](https://github.com/caseyjhol/node-red-contrib-tplink)

[![Github All Releases](https://img.shields.io/github/downloads/caseyjhol/node-red-contrib-tplink/total.svg?style=flat-square)](https://github.com/caseyjhol/node-red-contrib-tplink/releases)

[![npm](https://img.shields.io/npm/l/node-red-contrib-tplink.svg?style=flat-square)](https://github.com/caseyjhol/node-red-contrib-tplink/blob/master/LICENSE) [![GitHub contributors](https://img.shields.io/github/contributors/caseyjhol/node-red-contrib-tplink.svg?style=flat-square)](https://github.com/caseyjhol/node-red-contrib-tplink/graphs/contributors) [![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat-square)](https://github.com/Felixls/node-red-contrib-tplink-smarthome/issues)

This is a [Node-RED](https://nodered.org/) node that allows you to control smart plugs, switches, and bulbs from the TP-Link Kasa Smart ecosystem.

Under the hood, this uses [TP-Link Smart Home API](https://github.com/plasticrake/tplink-smarthome-api).

## Installation

Run the following command in the root directory of your Node-RED install

`$ npm install node-red-contrib-tplink`

or you can use the Palette Manager in Node-RED.

## Parameters

`Name` - Type in the name of the host manually or keep the default device name

`Device IP` - Type in the Device IP address manually or press the button to retrieve all locally available plug devices. To specify a plug in a multi-plug device, append a `/` followed by the plug number (zero-indexed). Use is optional. Alternatively, or additionally, an input message can include `msg.topic` with a device IP (and optional plug number). Examples: `192.168.1.101` or `192.168.1.101/3`

`Connection poll interval` - Interval that is used to poll availability of devices _(>500ms / Recommended 5000-10000ms)_. Set to `0` to disable availability checks (in which case, `OnlineEvents` will not trigger, even if started).

`Event poll interval` - Interval that is used to poll active devices for events _(>500ms / Recommended 1000-3000ms)_. Set to `0` to disable event polling (in which case, all events except OnlineEvents will not trigger, even if started).

`Output payload` - Types are `info`, `none`, `string`, `number`, `boolean`, `JSON`, and `timestamp`. Default type is `info` with a value of `getInfo`. If set to `info`, the selected payload command will be fired after completion, and the payload set accordingly. Output payload is ignored if the node input is a command or an event, in which case see below for more information

`Debug` - If enabled, will output device information to the flow editor debug tab.

## Inputs

Send in a message to control, command, or start/stop events.

- `topic` - Optional. The device IP (and optional plug). Any message without a topic will use the device configured via the parameters. If no topic is included, and no device is configured, the message will be ignored.

- `payload` - Required. Either a **control**, **command**, or **event action**. See below for details of each.

### Controls

Control a device by setting its properties.

> string | Object

-  `true` | `on` - Turn on the device
-  `false` | `off` - Turn off the device
-  `toggle` - Switch the power state of the device.
- Or as an object, all properties optional:
	-  `state: true` | `on` | `false` | `off` - Set device on or off
	-  `brightness: [1-100]` - Set brightness, if supported
	-  `temperature: [2700-6500]` - Set brightness (in kelvin), if supported
	-  `hsb: {hue, saturation, brightness}` - Set the color, if supported
	-  `led: true` | `false` - Turn the LED on or off, if supported

### Commands

> string

- `getInfo` - Fetch the device information.
- `getCloudInfo` - Fetch the device information from the cloud.
- `getQuickInfo` - Fetch most popular proprieties, such as username, device consumption, mode, lighting state, and many more. Supports multi-plug.
- `getMeterInfo` - Fetch the current device consumption. Supports multi-plug.
- `eraseStats` - Clear all the meter statistics. Supports multi-plug.

### Events

> string | array | Object

- `startMeterEvents`/`stopMeterEvents` - Subscribe to meter information events. Event emits on event polling interval.
- `startInfoEvents`/`stopInfoEvents` - Subscribe to information events. Event emits on event polling interval.
- `startPowerEvents`/`stopPowerEvents` - Subscribe to power on/off events. Event emits on device/plug change.
- `startPowerUpdateEvents`/`stopPowerUpdateEvents` - Subscribe to power on/off events. Event emits on event polling interval.
- `startInUseEvents`/`stopInUseEvents` - Subscribe to device usage events. Event emits on device/plug change.
- `startInUseUpdateEvents`/`stopInUseUpdateEvents` - Subscribe to device usage events. Event emits on event polling interval.
- `startOnlineEvents`/`stopOnlineEvents` - Subscribe to online/offline events. Event emits on poll interval.
- `stopAllEvents` - Unsubscribe all events.

_Multiple event actions can be sent at once, either as an array or as a string separated by "|". Alternatively, an array or string can be sent in the `events` property of an object._
Examples:

```js
['startMeterUpdateEvents', 'stopPowerEvents']
```

or

```js
{ events: 'startMeterUpdateEvents|stopPowerEvents' }
```

## For developers

This repo. is _(mainly)_ coded on [Node 10.3.0](https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V10.md#10.3.0) with [Node-RED 0.18.7](https://github.com/node-red/node-red/blob/master/CHANGELOG.md) on [Windows 10 Home Build 17134.81](https://support.microsoft.com/ro-ro/help/4100403/windows-10-update-kb4100403).

Runs succesfully in a [Raspberry Pi 3 Model B+](https://www.raspberrypi.org/products/raspberry-pi-3-model-b-plus/) on standard Raspbian Stretch's [Node 0.10.29](https://nodejs.org/en/blog/release/v0.10.29/) and matching Node-RED.

[![https://nodei.co/npm/node-red-contrib-tplink.png?downloads=true&downloadRank=true&stars=true](https://nodei.co/npm/node-red-contrib-tplink.png?downloads=true&downloadRank=true&stars=true)](https://www.npmjs.com/package/node-red-contrib-tplink)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

Bear in mind that the [GIT version](https://github.com/caseyjhol/node-red-contrib-tplink) *(source-code)* is always bleeding edge and may *(in most cases, will)* have changes that won't reflect in NPM/[release](https://github.com/caseyjhol/node-red-contrib-tplink/releases) version.

## 1.0.0-alpha.0 (2020-07-08)

### Changed

- Upgraded to tplink-smarthome-api@3.0.0-beta
- smart-plug and smart-light combined into one node called "kasa"
- getXxxxEvents changed to startXxxxEvents
- Events output "XxxxEvents" on msg.event instead of "getXxxxEvents"
- clearEvents changed to stopAllEvents
- Node status reflects number of devices connected if more than one

### Added

- stopXxxxEvents to match the startXxxxEvents
- Multiple devices allowable per node (using msg.topic on input)
- Individual plugs on multi-plug devices can be accessed
- Polling can be turned off if not needed

## 0.4.3 (2020-06-08)

### Fixed

- getMeterInfo no longer throws an error
- getInfoEvents works again (broken in v0.3.0)
- getMeterEvents works again (broken in v0.3.0)

### Added

- Debug property to help debug issues with devices

## 0.4.2 (2020-06-05)

### Fixed

- Set payload defaults to ensure same functionality as older versions
- Fix error when passing a boolean as payload
- Fix error catching if an error is encountered while running a command

## 0.4.1 (2020-06-04)

### Fixed

- Set payload defaults to ensure same functionality as older versions

## 0.4.0 (2020-06-04)

### Added

- Errors are now catchable
- Payload property can now be set - defaults to getInfo, so it's the same as in earlier versions

### Fixed

- Support using `'|'` for passing events again
- Support setting state to switch again

## 0.3.0 (2020-06-03)

### Added

- Hue, Saturation, and Brightness (HSB) support for multicolor bulbs
- HS300 (multi-plug) support

### Changed

- Separated getPowerEvents/getInUseEvents from getPowerUpdateEvents/getInUseUpdateEvents

### Fixed

- Fix emission of online/offline events in smart-plug node

## 0.2.1

### Changed

- Resized SVG icon
- Default name when adding a new device now includes its alias (e.g. Bedroom Light - HS220)

### Fixed

- Included SVG icon in package.json to ensure it is visible

## 0.2.0

### Added

- HS220 brightness support
- Support for turning off LED status light on switches

### Changed

- Payload can now be a JSON object to allow multiple settings to be changed at once
- Updated tplink-smarthome-api to v2
- Removed separate compiled files (moved files into nodes folder)
- Use SVG for icon instead of PNG

### Fixed

- Fix metering messages for plugs

### Breaking

- npm minimum version changed to v3.0.0
- node minimum version changed to v10.0.0

## 0.1.8 - Unreleased

### Added

- This file.
- Get most common info with **`getQuickInfo`** - *such as cloud account, power state, power consumption,  schedule, and many more*.
- **Filetree** hardcoded in `package.json` *(GIT builds clean now)*.
- Google **compiled code** in main `.js` files and formatted code in `.bak.js` files *- harder, better, faster, stronger*.

### Changed

- Fixed some typos *( @rajamalw )*.
- Commands sanity for `brightness` and `temperature` *- won't trigger unless device supports it ( @rajamalw )*.
- Updated documentation.

### Removed

Nothing but bad vibes.

nodes/kasa.js
const isPlainObject = obj => Object.prototype.toString.call(obj) === '[object Object]'

module.exports = function (RED) {
  'use strict'
  const { Client } = require('tplink-smarthome-api')

  function TpLinkNode(config) {
    RED.nodes.createNode(this, config)
    this.config = {
      name: config.name,
      deviceId: config.device,
      interval: parseInt(config.interval),
      eventInterval: parseInt(config.eventInterval),
      payload: config.payload === undefined ? 'getInfo' : config.payload,
      payloadType: config.payloadType === undefined ? 'info' : config.payloadType,
      debug: config.debug
    }

    // Only one client instance used per node
    const client = new Client({
      logLevel: config.debug ? 'debug' : 'warn',
      defaultSendOptions: {
        transport: 'tcp',
        timeout: this.config.eventInterval || 30000
      }
    })
    const moment = require('moment')
    const numeral = require('numeral')
    const context = this.context()

    const node = this
    node.client = client
    // Devices array will hold the device object returned
    // by client.getDevice(). Each device has all methods
    // of the underlying TP-Link API, including an event bus.
    node.devices = new Map()

    // For each new device in this node, a connection is established
    // using the client instance for this node. The returned
    // device object is added to the dictionary of devices, keyed
    // by the shortId (format <IP> or <IP>/<PLUG>). Once added,
    // event proxies are setup, and then polling begins.
    node.connectDevice = async function (id) {
      if (!node.devices.has(id)) {
        // Device shell/placeholder should be setup prior to calling this function
        return
      }

      let shellDevice = node.devices.get(id)
      if (!shellDevice.placeholder) {
        // This shouldn't ever happen - but just in case
        node.error('Attempting to connect to device that already has a connection in this node')
        return shellDevice
      }
      shellDevice.connecting = true

      let [deviceIP, plug] = id.split('/')
      let options = { host: deviceIP }
      if (plug) options.childId = plug

      try {
        let device = await client.getDevice(options)
        await device.getInfo()
        node.setupDevice(id, device)
        return device
      } catch (err) {
        shellDevice.connecting = false
        if (node.config.debug) node.error(`Error connecting to device ${id}: ${err}`)
        // Unable to connect initially. If device is later seen during discovery,
        // it will use the discovery response to 'connect' and everything will
        // then work normally.
        return
      }
    }

    node.setupDevice = function (id, device) {
      device.shortId = id
      device.events = []
      device.online = true

      if (node.config.eventInterval) {
        node.setupEventProxies(device)
        device.startPolling(node.config.eventInterval)
      }

      let queue = node.devices.has(id) && node.devices.get(id).queue
      node.devices.set(id, device)
      if (queue) queue.forEach(msg => node.processInput(msg, device))

      node.updateStatus()

      return device
    }

    // This section uses a single discovery poll to listen to all
    // devices. If the device is tracked in this node, a respective
    // event will be triggered. Note: No message will output if
    // the OnlineEvents have not been started (see below).
    // This is essentially a "isAlive polling" check for all devices.
    // This also catches any devices that failed to connect initially.
    if (node.config.interval)
      client.startDiscovery({
        discoveryInterval: node.config.interval,
        offlineTolerance: 1,
        breakoutChildren: false
      })

    const emitOnlineEvents = state => dev => {
      let host = dev.host
      let ids = [host]
      if (dev.children && !dev.childId)
        dev.children.forEach((x, childId) => ids.push(`${host}/${+childId.slice(-2)}`))

      ids.forEach(id => {
        if (node.devices.has(id)) {
          let device = node.devices.get(id)
          if (device.placeholder && !device.connecting && state == true) {
            // Probably initial connection to device failed. Now
            // that we have the device, set it up, then proceed.
            device = node.setupDevice(id, dev)
          }
          if (!device.placeholder) {
            device.online = state
            device.emit('OnlineEvents', { online: state, state })
          }
        }
      })
      node.updateStatus()
    }
    client.on('device-new', emitOnlineEvents(true))
    client.on('device-online', emitOnlineEvents(true))
    client.on('device-offline', emitOnlineEvents(false))

    client.on('error', error => node.error(`Underlying API error: ${error}`))

    // On input message, determine which device to reference. If a topic
    // is sent, that is used as the deviceId. When topic is omitted, use
    // the node configuration. If the device has not been connected to
    // yet, setup connection first then processing message.
    node.on('input', function (msg) {
      let shortId = msg.topic || node.config.deviceId

      if (!shortId) return

      if (node.devices.has(shortId)) {
        let device = node.devices.get(shortId)
        if (device.placeholder) {
          device.queue.push(msg)
        } else {
          node.processInput(msg, device)
        }
      } else {
        node.devices.set(shortId, {
          placeholder: true,
          queue: [msg]
        })
        node.connectDevice(shortId)
      }
    })

    // Input messages could be one of three categories, based
    // on the type of payload and data within. Once determined,
    // the message is routed onwards.
    // The three categories and their 'next' functions are:
    //   1. Control  >>  (Request made here)  >>  node.sendControlResult()
    //   2. Command  >>  node.handleCommand()
    //   3. Event action  >>  node.handleEventAction()
    node.processInput = function (msg, device) {
      if (node.config.debug)
        node.warn(`Processing input for device ${device.shortId}: ${JSON.stringify(msg)}`)

      let input = msg.payload
      let promises = []

      // OBJECT
      if (isPlainObject(input)) {
        // payload.state
        if (input.hasOwnProperty('state')) {
          if (/^(TOGGLE|SWITCH)$/i.test(input.state)) {
            promises.push(device.togglePowerState())
          } else if (/^(ON|TRUE)$/i.test(input.state)) {
            promises.push(device.setPowerState(true))
          } else if (/^(OFF|FALSE)$/i.test(input.state)) {
            promises.push(device.setPowerState(false))
          } else node.error('Invalid state value; Should be true|on|false|off')
        }

        // payload.brightness
        if (
          input.hasOwnProperty('brightness') &&
          (device.supportsBrightness || device.supportsDimmer) &&
          validateNumber('brightness', input.brightness, 1, 100)
        ) {
          if (device.deviceType == 'bulb') {
            promises.push(device.lighting.setLightState({ brightness: input.brightness }))
          } else {
            promises.push(device.dimmer.setBrightness(input.brightness))
          }
        }

        // payload.temperature
        if (
          input.hasOwnProperty('temperature') &&
          device.supportsColorTemperature &&
          validateNumber('temperature', input.temperature, 2700, 6500)
        ) {
          promises.push(device.lighting.setLightState({ color_temp: input.temperature }))
        }

        // payload.hsb
        if (
          input.hasOwnProperty('hsb') &&
          device.supportsColor &&
          validateNumber('hue', input.hsb.hue, 0, 360) &&
          validateNumber('saturation', input.hsb.saturation, 0, 100) &&
          validateNumber('brightness', input.hsb.brightness, 1, 100)
        ) {
          promises.push(device.lighting.setLightState({ ...input.hsb }))
        }

        // payload.led
        if (input.hasOwnProperty('led') && device.hasOwnProperty('setLedState')) {
          promises.push(device.setLedState(input.led))
        }

        // payload.events
        if (input.hasOwnProperty('events')) {
          if (typeof input.events === 'array' || input.events instanceof Array) {
            input.events.forEach(event => node.handleEventAction(device, event))
          } else if (typeof input.events === 'string') {
            input.events.split('|').forEach(event => node.handleEventAction(device, event))
          }
        }

        // ARRAY
      } else if (typeof input === 'array' || input instanceof Array) {
        input.forEach(event => node.handleEventAction(device, event))

        // STRING
      } else if (typeof input === 'string' || input instanceof String) {
        let payload = input

        if (/^(TOGGLE|SWITCH)$/i.test(payload)) {
          promises.push(device.togglePowerState())
        } else if (/^(ON|TRUE)$/i.test(payload)) {
          promises.push(device.setPowerState(true))
        } else if (/^(OFF|FALSE)$/i.test(payload)) {
          promises.push(device.setPowerState(false))
        } else if (input.includes('|') || /^(START|STOP)/i.test(payload)) {
          input.split('|').forEach(event => node.handleEventAction(device, event))
        } else {
          node.handleCommand(device, input)
        }

        // BOOLEAN
      } else if (input === true || input === false) {
        promises.push(device.setPowerState(input))

        // UNSUPPORTED
      } else {
        node.error(`Invalid input: ${JSON.stringify(input)}`)
      }

      if (promises.length)
        Promise.all(promises)
          .then(() => node.sendControlResult(device, input))
          .catch(error => {
            if (node.config.debug) node.error(`Error controlling device: ${error}`)
          })
    }

    // Device state was already updated (ie the device was
    // controlled). Here we send an output message based
    // on the node configuration.
    node.sendControlResult = function (device, inputPayload) {
      let msg = {}

      switch (node.config.payloadType) {
        case 'json':
          msg.payload = JSON.parse(inputPayload)
          break
        case 'date':
          msg.payload = Date.now()
          break
        case 'bool':
          msg.payload = inputPayload === 'true'
          break
        case 'num':
          msg.payload = parseInt(inputPayload)
          break
        case 'info':
          node.handleCommand(device, node.config.payload)
          return
      }

      msg.topic = device.shortId
      node.send(msg)
    }

    // Special commands handled here. Also, any input strings that
    // didn't match anything in the processor get sent here and will
    // error by default.
    node.handleCommand = function (device, cmd) {
      let promise
      switch (cmd) {
        case 'getInfo':
          promise = device.getSysInfo()
          break
        case 'getCloudInfo':
          promise = device.cloud.getInfo()
          break
        case 'getQuickInfo':
          promise = device.getInfo()
          break
        case 'getMeterInfo':
          promise = device.emeter.getRealtime()
          break
        case 'eraseStats':
          promise = device.emeter.eraseStats()
          break
        default:
          return node.error(`Invalid input: ${cmd}`)
      }

      promise
        .then(info =>
          node.send({
            topic: device.shortId,
            payload: {
              ...info,
              timestamp: moment().format()
            }
          })
        )
        .catch(node.error)
    }

    // For each device, an event proxy system is set up that maps the underlying
    // events to the node-based events on the same bus. Only those node-based events
    // that have been turned on (via node.handleEventAction) will end up sending
    // an output on the node.
    node.setupEventProxies = function (device) {
      let powerPrefix = device.deviceType == 'bulb' ? 'lightstate' : 'power'
      device.on(powerPrefix + '-on', () =>
        device.emit('PowerEvents', { powerOn: true, state: true })
      )
      device.on(powerPrefix + '-off', () =>
        device.emit('PowerEvents', { powerOn: false, state: false })
      )
      device.on(powerPrefix + '-update', powerOn => {
        device.emit('PowerUpdateEvents', { powerOn, state: powerOn })
        device.emit('InfoEvents') // placed here to take advantage of polling interval
      })

      device.on('in-use', () => device.emit('InUseEvents', { inUse: true, state: true }))
      device.on('not-in-use', () => device.emit('InUseEvents', { inUse: false, state: false }))
      device.on('in-use-update', inUse => device.emit('InUseUpdateEvents', { inUse, state: inUse }))

      device.on('emeter-realtime-update', emeter => {
        device.emit('MeterEvents', { emeter })
        if (node.devices.size == 1) node.updateStatus()
      })

      device.on('polling-error', error => {
        if (node.config.debug) node.error(`Polling error for ${device.shortId}: ${error}`)
      })
    }

    // Based on the input command, a node-based event is turned on or off as requested.
    node.handleEventAction = function (device, cmd) {
      if (cmd === 'stopAllEvents') return device.removeAllListeners()

      let [match, startStop, event] = cmd.match(/^(start|stop)(.+)/)

      if (!match) return

      if (startStop === 'start' && !device.eventNames().includes(event)) {
        device.on(event, node.makeEventHandler(device, event))
      } else if (startStop === 'stop') {
        device.removeAllListeners(event)
      }
    }

    // When a node-based event is turned on, this generates the callback function
    // that takes the passed properties from the underlying event (based on the
    // parameters in node.setupEventProxies) and outputs a formatted message.
    node.makeEventHandler = function (device, event) {
      if (event == 'InfoEvents')
        return () =>
          node.send({
            topic: device.shortId,
            payload: device.sysInfo
          })

      return passedProps =>
        node.send({
          topic: device.shortId,
          payload: {
            event: event,
            timestamp: moment().format(),
            ...passedProps
          }
        })
    }

    // The node status is different, depending on how many devices
    // are using this node. If only one device is connected then
    // the cached info for that device is used. Any more than one
    // device, and only the number of devices is used. Number of
    // offline device is always shown if more than 0.
    node.updateStatus = function () {
      let onlineDevice
      let numOnline = 0
      let numOffline = 0
      node.devices.forEach(dev => {
        if (dev.online) {
          numOnline++
          onlineDevice = dev
        } else numOffline++
      })

      let status = {
        fill: 'green',
        shape: 'dot'
      }

      if (numOnline === 0) {
        status.fill = 'red'
        status.shape = 'ring'
        status.text = 'No devices connected'
      } else if (numOnline === 1) {
        let device = onlineDevice
        let state = device.relayState ? 'ON' : 'OFF'
        if (device.supportsEmeter) {
          let emeter = device.emeter.realtime
          state += ` [${numeral(emeter.power_mw / 1000.0).format('0.[00]')}W: ${numeral(
            emeter.voltage_mv / 1000.0
          ).format('0.[0]')}V@${numeral(emeter.current_ma / 1000.0).format('0.[000]')}A]`
        }
        status.text = `One device connected (${state})`
      } else {
        status.text = `${numOnline} devices connected`
      }

      if (numOffline > 0) {
        status.fill = 'yellow'
        status.text += ` (${numOffline} offline)`
      }

      node.status(status)
    }

    if (this.config.deviceId) {
      // If configured with a device, initiate connection
      node.devices.set(this.config.deviceId, {
        placeholder: true,
        queue: []
      })
      node.connectDevice(this.config.deviceId)
    } else {
      // Otherwise, initialize status (with no devices)
      node.updateStatus()
    }

    // Helpers
    function validateNumber(propName, value, min, max) {
      if ((value >= min && value <= max) || value === undefined) {
        return true
      } else {
        node.error(`Invalid ${propName} value; Should be between ${min} and ${max}`)
        return false
      }
    }

    // Cleanup when node destroyed
    node.on('close', function () {
      // Important: All code in this function must be sure
      // to not cause any errors. If any error is uncaught,
      // cleanup may not finish properly and there will
      // be a memory leak!
      node.devices.forEach(device => {
        device.stopPolling && device.stopPolling()
        device.closeConnection && device.closeConnection()
      })
      node.client.stopDiscovery()
    })
  }

  //Make available as node
  RED.nodes.registerType('kasa', TpLinkNode)

  RED.httpAdmin.get('/smarthome/devices', (req, res) => {
    try {
      const client = new Client()
      let discoveryTimeout = 10000
      let devices = []
      client.on('device-new', device => {
        devices.push({
          host: device.host,
          alias: device.alias,
          childId: device.childId
        })
      })
      client.startDiscovery()
      setTimeout(() => {
        client.stopDiscovery()
        res.end(JSON.stringify(devices))
      }, discoveryTimeout)
    } catch (error) {
      res.sendStatus(500).send(error.message)
    }
  })
  RED.httpAdmin.get('/smarthome/device', (req, res) => {
    if (!req.query.host) return res.status(500).send('Missing Device IP…')
    const client = new Client()
    client
      .getDevice(req.query)
      .then(device => {
        res.end(
          JSON.stringify({
            model: device.model,
            alias: device.alias
          })
        )
      })
      .catch(error => {
        res.sendStatus(500).send(error.message)
      })
  })
}
